docs/proposals/spiffe-spire-integration.rst

```diff
@@ -0,0 +1,132 @@
+Spire and it's stack.
+
+Changes
+-------
+
+Spire works to a server/slave architecture. As such, the Vault HA cluster will
+be replaced with a Spire HA server cluster. This means the Terraform Vault
+provider will also be replaced with a Spire equivalent.
+
+Spire agents will then sit on each Kubernetes instance that are able to connect
+to the Spire server using AWS identifiers. These agents, once authenticated with
+the Spire server cluster are then able to set up PKI within the instances
+through use of attestor plugins with suitable selectors to authenticate
+identity. A spire-helper type tool will need to be developed to setup PKI, just
+like how vault-helper manages this.
+
+Steps as I currently see them (Subject to Change):
+
+- Set up Spire HA server cluster.
+- Create Spire Terraform provider.
+- Authenticate Spire agents on AWS instances against the server cluster.
+- Create spire-helper type tool.
+- Set up PKI with Kubernetes components using Spire agents.
+
+Notable items
+-------------
+
+- This represents a significant amount of implementation time.
+- Spire is still in an early stage meaning it hasn't been as industry tested in
+  production like Vault has been.
+- Spire is still in very much heavy development so breaking changes are highly
+  probable.
+- It may just end up not working and be incompatible with Tarmak clusters.
+
+Out of scope
+------------
+
+Avoiding writing our own plugins. This is better left to experts and if is
+required, Spire should be dropped due to not being ready yet for our
+requirements.
+
+Testing Review
+--------------
+I was able to create Spire servers on each Vault instance. These all connect to
+a single PostgreSQL elected master as a backend. PostgreSQL replicas reside on
+each Vault instance that promote a leader through Consul.
+
+Spire agents reside on each Kubernetes instance that are able to attest through
+their AWS identity meta data. No token is required, they can simply connect to a
+spire sever and it's done!
+
+SVIDs can be generated for each Kubernetes component that needs them through
+first registering components with kernel primitive selectors (uid, gid) against the
+server(s). The components can then request their SVID against the agent and receive
+their key and certificate material.
+
+Problems During Testing
+-----------------------
+
+Spire is not currently well built for HA and really only expects a single server
+in cluster. PostgreSQL support for HA is quite poor from what I have found,
+whilst also being the only supported "HA" backend. I was able to use a tool Patroni_ that
+facilitates a PostgreSQL cluster with Consul however only the master is in write
+mode. All replicas on other instances are in read mode which causes spire severs
+on those instances to panic. Spire servers have to connect to the PostgreSQL
+cluster master elected through Consul.
+
+Spire uses Elliptic Curve keys which is odd and can be annoying.
```

Comment:
Not too sure either provide more evidence or get rid of this sentence. There are good reasons IMHO for ECC (https://blog.cloudflare.com/ecdsa-the-digital-signature-algorithm-of-a-better-internet/)